wagtail/core/blocks/struct_block.py

```diff
@@ -149,6 +149,45 @@ def to_python(self, value):
             for name, child_block in self.child_blocks.items()
         ])
 
+    def bulk_to_python(self, values):
+        # values is a list of dicts; split this into a series of per-subfield lists so that we can
+        # call bulk_to_python on each subfield
+
+        values_by_subfield = {}
+        for name, child_block in self.child_blocks.items():
+            # We need to keep track of which dicts actually have an item for this field, as missing
+            # values will be populated with child_block.get_default(); this is expected to be a
+            # value in the block's native type, and should therefore not undergo conversion via
+            # bulk_to_python.
+            indexes = []
+            raw_values = []
+            for i, val in enumerate(values):
+                if name in val:
+                    indexes.append(i)
+                    raw_values.append(val[name])
+
+            converted_values = child_block.bulk_to_python(raw_values)
+            # create a mapping from original index to converted value
+            converted_values_by_index = dict(zip(indexes, converted_values))
+
+            # now loop over all list indexes, falling back on the default for any indexes not in
+            # the mapping, to arrive at the final list for this subfield
+            default = child_block.get_default()
```

Comment:
Good point! I'm pretty sure that all code within StreamField itself treats the data as immutable and will always rebuild the full data structure when making changes rather than modifying in place, but I'm sure there must be user code out there which does things like `.append` on a ListBlock value - and that'll end up modifying every record that was created through this code path.

It looks like in addition to fixing the code here to make a distinct call to `get_default` on every iteration, we also need to fix ListBlock.get_default so that it doesn't return the same instance every time. (StructBlock and StreamBlock already take care of this, because they have `get_default` methods that cast `self.meta.default` to the expected data type)

Reply:
Now fixed in https://github.com/wagtail/wagtail/pull/5976/commits/bc175c73a4f9f25d09b4ae1b3a892572755d4945.